[DCJ-625] unlockDataset endpoint also unlocks any load tags

okotsopoulos opened the pull request:
__Jira ticket__: https://broadworkbench.atlassian.net/browse/DCJ-625

## Addresses

When a dataset steward initiates a manual unlock of their dataset via a call to https://data.terra.bio/swagger-ui.html#/datasets/unlockDataset, their intention is to clear any stuck locks on the dataset which are impeding its availability.

Some flights which lock datasets also lock a *load tag* -- flights which ingest files to a dataset are our most notable example.  So if a dataset has a stuck lock, sometimes a load tag also remains locked within that dataset.

If a user wanted to reuse the same load tag for operations on the dataset, they'd be unable to do so, even if the dataset itself has been unlocked.

## Summary of changes

Now that load tag locks are scoped to the targeted dataset as of https://github.com/DataBiosphere/jade-data-repo/pull/1785, this PR expands `DatasetUnlockFlight` to include new step `LoadManualUnlockStep`, which removes any stuck lock held by the user-supplied `lockName` on the targeted dataset (we don't care about what specific load tag may or may not be locked).

The original implementation expanded existing `LoadUnlockStep` to support this use case. In this case, two very simple steps that share service code is preferable for maintainability and comprehension over one overloaded step.

## Testing Strategy

Expanded unit tests.

<!-- Reminder -->
<!-- Two CODEOWNERS will be automatically assigned to review this pull request. If you otherwise have two reviewers, you do not need to wait for their review. -->

okotsopoulos committed:
LoadDao.unlockLoad treats load tag specification as optional

When clearing a stuck lock on a dataset, we also want to clear any locks stuck on load tags in the dataset.
When no load tag is specified when calling LoadDao.unlockLoad, delete any rows matching the dataset and locking flight ID.
Otherwise, delete any rows matching the dataset, load tag, and locking flight ID.

okotsopoulos committed:
Add new LoadService.unlockLoad method to clear lock on dataset

Without needing to specify the load tag.

okotsopoulos committed:
LoadUnlockStep accepts user-supplied lockName

The default behavior when unspecified is to unlock the current flight's ID (when clearing a lock taken out earlier in the flight).
In the case of manual dataset unlocking, we want to clear a lock taken out by a past flight.
Manual dataset unlocking will not have the load tag to unlock in its flight context, so unlocks any load tag associated with the dataset ID and lock name.

okotsopoulos committed:
DatasetUnlockFlight includes LoadUnlockStep

The lock name from the UnlockResourceRequest is passed to the step constructor.

okotsopoulos committed:
Add a unit test for DatasetService.manualUnlock

Ensures that the dataset ID is supplied as a flight input parameter.

okotsopoulos committed:
Merge branch 'develop' into okotsopo-DCJ-625-unlock-dataset-and-load-tag

okotsopoulos committed:
Expand unlockDataset Swagger endpoint description

okotsopoulos committed:
Sonar: remove @Transactional annotation from non-public method

LoadDao.lookupLoadLocks only exists for unit testing.

rushtong reviewed (approved):
This looks reasonable to me, thank you 👍🏽

pshapiro4broad commented on src/main/java/bio/terra/service/load/flight/LoadUnlockStep.java:
```diff
@@ -1,33 +1,78 @@
    *
    * <ul>
-   *   <li>{@link LoadMapKeys#LOAD_TAG} in the input parameters or working map
+   *   <li>(Optional) {@link LoadMapKeys#LOAD_TAG} in the input parameters or working map (if
+   *       unavailable, any lock associated with the dataset and {@code lockName} will be cleared)
```
I think conditional behavior like this makes it harder to maintain steps. Would it be hard to have owning flight always provide the lock name in `LoadMapKeys.LOAD_TAG`? It looks like the only case affected by that change would be `DatasetUnlockFlight`, which is only called from a single place.

pshapiro4broad commented on src/main/java/bio/terra/service/load/flight/LoadUnlockStep.java:
```diff
@@ -1,33 +1,78 @@
   public StepResult doStep(FlightContext context) {
-    LoadLockKey loadLockKey = loadService.getLoadLockKey(context);
-    loadService.unlockLoad(loadLockKey, context.getFlightId());
+    String lockName = getLockName(context);
+    try {
+      LoadLockKey loadLockKey = loadService.getLoadLockKey(context);
+      loadService.unlockLoad(loadLockKey, lockName);
+    } catch (LoadLockFailureException ex) {
+      // The flight context may not have a load tag.  We'll clear any lock associated with the
+      // flight's dataset and lock name in that case.
+      loadService.unlockLoad(IngestUtils.getDatasetId(context), lockName);
+    }
```
I think it would be clearer to only make a single call to `unlockLoad()`. That would also make it clear that the `catch` only applies to `getLoadLockKey()` and not to `unlockLoad()`. Something like this would work:
```suggestion
    LoadLockKey loadLockKey;
    try {
       loadLockKey = loadService.getLoadLockKey(context);
    } catch (LoadLockFailureException ex) {
      // The flight context may not have a load tag.  We'll clear any lock associated with the
      // flight's dataset and lock name in that case.
      loadLockKey = new LoadLockKey(null, IngestUtils.getDatasetId(context));
    }
    loadService.unlockLoad(loadLockKey, getLockName(context));
```

You can add another constructor to `LoadLockKey()` that supports only the dataset ID to avoid having to pass `null` here.

okotsopoulos replied:
Moot point now that I've broken up into a separate step (though I did add an alt `LoadLockKey` constructor), thanks again for the suggestion!

pshapiro4broad commented on src/main/java/bio/terra/service/load/flight/LoadUnlockStep.java:
```diff
@@ -1,33 +1,78 @@
   }
+
+  /**
+   * @return the name of the lock to remove (if unspecified, defaults to this flight's ID)
```
I worry about the complexity of making steps work in different ways based on how they're used. Another approach would be to make this a required flight context property, but that would require adding an extra step to the flight to create the data this step needs.

okotsopoulos replied:
Good point -- now that this is in its own step, we have two simple steps rather than one complex one, while still reusing the service code.

pshapiro4broad reviewed (approved):
This looks OK but I worry about making steps too flexible. If most of the code is in the service layer, maybe a different step for different contexts would be easier to maintain.

pshapiro4broad commented on src/test/java/bio/terra/service/load/LoadDaoUnitTest.java:
```diff
@@ -161,4 +162,39 @@ private void testLoadCandidates(
+    final String flightX = "flightIdX";
+    final String flightY = "flightIdY";
+    final UUID datasetId = dataset.getId();
+    final LoadLockKey loadTag1_dataset = new LoadLockKey(loadTag1, datasetId);
+    final LoadLockKey loadTag2_dataset = new LoadLockKey(loadTag2, datasetId);
```
Sonar flags these names as not conforming to our code standard. It's not a huge deal but it would be nice to avoid adding sonar warnings if possible
```suggestion
    final LoadLockKey loadTag1Dataset = new LoadLockKey(loadTag1, datasetId);
    final LoadLockKey loadTag2Dataset = new LoadLockKey(loadTag2, datasetId);
```

okotsopoulos commented:
> This looks OK but I worry about making steps too flexible. If most of the code is in the service layer, maybe a different step for different contexts would be easier to maintain.

Good idea @pshapiro4broad  -- I'll break this out into a new step.  Then we'll have two very simple steps that share service code, rather than one complex one.

okotsopoulos committed:
Add alt LoadLockKey constructor (no load tag)

okotsopoulos committed:
Apply Sonar's variable naming convention

okotsopoulos committed:
New step LoadManualUnlockStep in DatasetUnlockFlight

Allows me to restore LoadUnlockStep into its simpler prior state.
Now we have two very simple steps which share LoadService code, rather than one more complicated step whose behavior is less clear.
